<a href="https://colab.research.google.com/github/max-moore/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2-make-features/LS_DS_112_Make_Features.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 2*

---

# Make Features 

- Student should be able to understand the purpose of feature engineering
- Student should be able to work with strings in pandas
- Student should be able to work with dates and times in pandas
- Student should be able to modify or create columns of a dataframe using the `.apply()` function


Helpful Links:
- [Minimally Sufficient Pandas](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)
- [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)
- Python Data Science Handbook
  - [Chapter 3.10](https://jakevdp.github.io/PythonDataScienceHandbook/03.10-working-with-strings.html), Vectorized String Operations
  - [Chapter 3.11](https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html), Working with Time Series
- [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)

# [Objective](#feature-engineering) - The Purpose of Feature Engineering



## Overview

Feature Engineering is the process of using a combination of domain knowledge, creativity and the pre-existing columns of a dataset to create completely new columns.

 Machine Learning models try to detect patterns in the data and then associate those patterns with certain predictions. The hope is that by creating new columns on our dataset that we can expose our model to new patterns in the data so that it can make better and better predictions.

This is largely a matter of understanding how to work with individual columns of a dataframe with Pandas --which is what we'll be practicing today!

## Follow Along

Columns of a dataframe hold each hold a specific type of data. Lets inspect some of the common datatypes found in datasets and then we'll make a new feature on a dataset using pre-existing columns.

In [88]:
import pandas as pd
import numpy as np

pd.set_option('display.max_rows', 150)
pd.set_option('display.max_columns', 100)

In [ ]:
# Lets take a look at the Ames Iowa Housing Dataset:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')

df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

### Specific Columns hold specific kinds of data

In [ ]:
df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
                  ...   
MoSold             int64
YrSold             int64
SaleType          object
SaleCondition     object
SalePrice          int64
Length: 81, dtype: object

Some columns hold integer values like the `BedroomAbvGr` which stands for "Bedrooms Above Grade." This is the number of non-basement bedrooms in the home.

For more information on specific column meanings view the [data dictionary](https://github.com/ryanleeallred/datasets/blob/master/Ames%20Housing%20Data/data_description.txt).

In [ ]:
# Look at the first ten rows of the `BedroomAbvGr` column.
# Looks like integers to me!
df['BedroomAbvGr']

Some columns hold float values like the `LotFrontage` column.

In [ ]:
# Look at the first ten rows of the `BedroomAbvGr` column.
df['LotFrontage']

Hmmm, do the values above look like floats to you?

They all have .0 on them so technically they're being stored as floats, but *should* they be stored as floats?

Lets see what all of the possible values for this column are.

In [ ]:
df['LotFrontage'].value_counts()

Looks to me like the `LotFrontage` column originally held integer values but was cast to a `float` meaning that each original integer values was converted to its corresponding float representation. 

Any guesses as to why that would have happened?


HINT: What's the most common `LotFrontage` value for this column?

In [ ]:
# NaN is the most common value in this column. What is a NaN
df['LotFrontage'].value_counts(dropna=False)

`NaN` stands stands for "Not a Number" and is the default missing value indicator with Pandas. This means there were cells in this column that didn't have a LotFrontage value recorded for those homes. 

This is where domain knowledge starts to come in. Think about the context we're working with here: houses. What might a null or blank cell representing "Linear feet of street connected to property" mean in the context of a housing dataset?

Ok, so maybe it makes seanse to have some NaNs in this column. What is the datatype of a NaN value?

Perhaps some of this data is truly missing or unrecorded data, but sometimes `NaNs` are more likely to indicate something that was "NA" or "Not Applicable" to a particular observation. There could be multiple reasons why there was no value recorded for a particular feature.

Remember, that Pandas tries to maintain a single datatype for all values in a column, and therefore...

In [ ]:
import numpy as np

# What is the datatype of NaN?
type(np.NaN)

The datatype of a NaN is float!  This means that if we have a column of integer values, but the column has even a single `NaN` that column will not be treated with the integer datatype but all of the integers will be converted to floats in order to try and preserve the same datatype throughout the entire column.

You can see already how understanding column datatypes is crucial to understanding how Pandas help us manage our data.

### Making new Features

Lets slim down the dataset and consider just a few specific columns:

- `TotalBsmtSF`
- `1stFlrSF`
- `2ndFlrSF`
- `SalePrice1`


In [ ]:
# I can make a smaller dataframe with a few specific column headers
# by passing a list of column headers inside of the square brackets
small_df = df[['TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'SalePrice']].copy()

small_df.head()

small_df.dtypes

### Syntax for creating new columns

When making a new column on a dataframe, we have to use the square bracket syntax of accessing a column. We can't use "dot syntax" here.

In [ ]:
# Lets add up all of the square footage to get a single square footage 
# column for the entire dataset

# Using bracket syntax to make a new 'TotalSquareFootage' column
small_df['TotalSquareFootage'] = small_df['TotalBsmtSF'] + small_df['1stFlrSF'] + small_df['2ndFlrSF']

small_df.head()

In [ ]:
# Lets make a nother new column that is 'PricePerSqFt' by
# dividing the price by the square footage

small_df['price_per_square_ft'] = small_df['SalePrice'] / small_df['TotalSquareFootage']

small_df.head()

Ok, we have made two new columns on our small dataset.

- What does a **high** `PricePerSqFt` say about a home that the square footage and price alone don't capture as directly?

- What does a **low** `PricePerSqFt` say about a home that the square footage and price alone don't directly capture?



## Challenge

I hope you can see how we have used existing columns to create a new column on a dataset that say something new about our unit of observation. This is what making new features (columns) on a dataset is all about and why it's so essential to data science --particularly predictive modeling "Machine Learning." 

We'll spend the rest of the lecture and assignment today trying to get as good as we can at manipulating (cleaning) and creating new columns on datasets.

# [Objective](#work-with-strings) Work with Strings with Pandas

## Overview

So far we have worked with numeric datatypes (ints and floats) but we haven't worked with any columns containing string values. We can't simply use arithmetic to manipulate string values, so we'll need to learn some more techniques in order to work with this datatype.

## Follow Along

We're going to import a new dataset here to work with. This dataset is from LendingClub and holds information about loans issued in Q4 of 2018. This dataset is a bit messy so it will give us plenty of opportunities to clean up existing columns as well as create new ones.

The `!wget` shell command being used here does exactly the same thing that your browser does when you type a URL in the address. It makes a request or "gets" the file at that address. However, in our case the file isn't a webpage, it's a compressed CSV file. 

Try copying and pasting the URL from below into your browser, did it start an automatic download? Any URLs like this that start automatic downloads when navigated to can be used along with the `!wget` command to bring files directly into your notebook's memory.

### Load a new dataset

In [ ]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-07-07 16:58:23--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 54.149.64.5, 52.11.197.48, 52.42.46.127, ...
Connecting to resources.lendingclub.com (resources.lendingclub.com)|54.149.64.5|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [       <=>          ]  22.21M  1.82MB/s    in 13s     

2020-07-07 16:58:37 (1.77 MB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [23289953]



We need to use the `!unzip` command to extract the csv from the zipped folder.

In [ ]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


We can also use bash/shell commands to look at the raw file using the `!head` and `!tail` commands

In [ ]:
!head LoanStats_2018Q4.csv

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [ ]:
!tail LoanStats_2018Q4.csv

As we look at the raw file itself, do you see anything that might cause us trouble as we read in the CSV file to a dataframe?

In [165]:
# Read in the CSV

df = pd.read_csv('LoanStats_2018Q4.csv')

df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0,1,2,3,4,7,13,18,19,24,25,27,28,29,30,31,32,34,36,37,38,39,40,41,42,43,44,46,49,50,51,53,54,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,113,114,115,116,117,118,119,120,121,126,127,131,132,134,135,136,141,142,143) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
id  member_id loan_amnt funded_amnt funded_amnt_inv term       int_rate installment grade sub_grade emp_title            emp_length home_ownership annual_inc verification_status issue_d  loan_status pymnt_plan url desc purpose            title                   zip_code addr_state dti   delinq_2yrs earliest_cr_line inq_last_6mths mths_since_last_delinq mths_since_last_record open_acc pub_rec revol_bal revol_util total_acc initial_list_status out_prncp out_prncp_inv total_pymnt     total_pymnt_inv total_rec_prncp total_rec_int total_rec_late_fee recoveries collection_recovery_fee last_pymnt_d last_pymnt_amnt next_pymnt_d last_credit_pull_d collections_12_mths_ex_med mths_since_last_major_derog policy_code application_type annual_inc_joint dti_joint verification_status_joint acc_now_delinq tot_coll_amt tot_cur_bal open_acc_6m open_act_il open_il_12m open_il_24m mths_since_rcnt_il total_bal_il il_util open_rv_12m open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m acc_open_past_24mths avg_cur_bal bc_open_to_buy bc_util chargeoff_within_12_mths delinq_amnt mo_sin_old_il_acct mo_sin_old_rev_tl_op mo_sin_rcnt_rev_tl_op mo_sin_rcnt_tl mort_acc mths_since_recent_bc mths_since_recent_bc_dlq mths_since_recent_inq mths_since_recent_revol_delinq num_accts_ever_120_pd num_actv_bc_tl num_actv_rev_tl num_bc_sats num_bc_tl num_il_tl num_op_rev_tl num_rev_accts num_rev_tl_bal_gt_0 num_sats num_tl_120dpd_2m num_tl_30dpd num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit total_il_high_credit_limit revol_bal_joint sec_app_earliest_cr_line sec_app_inq_last_6mths sec_app_mort_acc sec_app_open_acc sec_app_revol_util sec_app_open_act_il sec_app_num_rev_accts sec_app_chargeoff_within_12_mths sec_app_collections_12_mths_ex_med sec_app_mths_since_last_major_derog hardship_flag hardship_type hardship_reason hardship_status deferral_term hardship_amount hardship_start_date hardship_end_date payment_plan_start_date hardship_length hardship_dpd hardship_loan_status orig_projected_additional_accrued_interest hardship_payoff_balance_amount hardship_last_payment_amount debt_settlement_flag debt_settlement_flag_date settlement_status settlement_date settlement_amount settlement_percentage                                    settlement_term                              
NaN NaN       18000     18000       18000            36 months   8.81%  570.81      A     A5        Sever                < 1 year   RENT           40000      Source Verified     Dec-2018 Charged Off n          NaN NaN  debt_consolidation Debt consolidation      100xx    NY         12.81 0           Apr-2014         0              NaN                    NaN                    15       0       16147     35.2%      21        w                   0.00      0.00          6265.69         6265.69         5136.86         1128.83       0.0                0.0        0.0                     Oct-2019     570.81          NaN          May-2020           0                          NaN                         1           Individual       NaN              NaN       NaN                       0              0            28062       1           4           1           4           7                  11915        79      3           6           6591       46       45900            0      0           1            10                   1871        22177          38.1    0                        0           23                 56                   4                     4              0        4                    NaN                      8                     NaN                            0                     5              6               8           10        4         11            15            6                   15       0                0            0                  4                  100            12.5     

The extra rows at the top and bottom of the file have done two things:

1) The top row has made it so that the entire dataset is being interpreted as column headers

2) The bottom two rows have been read into the 'id' column and are causing every column to have at least two `NaN` values in it.

In [166]:
# We can fix the header problem by using the 'skiprows' parameter
df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1,skipfooter=2)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support skipfooter; you can avoid this warning by specifying engine='python'.
  


Lets look at the NaN values of each column so that you can see the problem that the extra rows at the bottom of the file are creating for us

In [91]:
# Sum null values by column and sort from least to greatest

df.isnull().sum().sort_values()

inq_fi                                             0
mo_sin_old_rev_tl_op                               0
delinq_amnt                                        0
chargeoff_within_12_mths                           0
acc_open_past_24mths                               0
inq_last_12m                                       0
total_cu_tl                                        0
total_rev_hi_lim                                   0
open_rv_24m                                        0
open_rv_12m                                        0
total_bal_il                                       0
open_il_24m                                        0
open_il_12m                                        0
open_act_il                                        0
open_acc_6m                                        0
tot_cur_bal                                        0
tot_coll_amt                                       0
acc_now_delinq                                     0
application_type                              

In [ ]:
# Address the extra NaNs in each column by skipping the footer as well.

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support skipfooter; you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


For good measure, we'll also drop some columns that are made up completely of NaN values.

Why might LendingClub have included columns in their dataset that are 100% blank?

In [167]:
df1 = df.drop(['desc', 'url', 'member_id', 'id'], axis=1)

df1.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0      18000        18000          18000.0   36 months    8.81%       570.81   
1      18500        18500          18500.0   60 months   12.98%       420.75   
2      32000        32000          32000.0   60 months   14.47%       752.41   
3       5000         5000           5000.0   36 months   17.97%       180.69   
4      13000        13000          13000.0   36 months   23.40%       505.95   

  grade sub_grade             emp_title emp_length home_ownership  annual_inc  \
0     A        A5                 Sever   < 1 year           RENT     40000.0   
1     B        B5  Supply Chain Analyst    5 years           RENT     96500.0   
2     C        C2               Teacher  10+ years       MORTGAGE    110000.0   
3     D        D1        Administrative    6 years       MORTGAGE     59280.0   
4     E        E1   Sale Representative    2 years       MORTGAGE     90000.0   

  verification_status   issue_d  loan_status pymnt_plan             purpose  \
0     Source Verified  Dec-2018  Charged Off          n  debt_consolidation   
1     Source Verified  Dec-2018      Current          n         credit_card   
2     Source Verified  Dec-2018      Current          n    home_improvement   
3     Source Verified  Dec-2018   Fully Paid          n  debt_consolidation   
4            Verified  Dec-2018      Current          n               other   

                     title zip_code addr_state    dti  delinq_2yrs  \
0       Debt consolidation    100xx         NY  12.81            0   
1  Credit card refinancing    554xx         MN  27.86            1   
2         Home improvement    152xx         PA  26.11            0   
3       Debt consolidation    490xx         MI  10.51            0   
4                    Other    191xx         PA  39.73            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0         Apr-2014               0                     NaN   
1         May-2009               0                     7.0   
2         Oct-2000               0                    24.0   
3         Apr-2011               0                     NaN   
4         May-2007               0                     NaN   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN        15        0      16147      35.2%         21   
1                     NaN        19        0      19000      62.1%         38   
2                     NaN        18        0      14833        43%         32   
3                     NaN         8        0       4599      19.1%         13   
4                     NaN        14        0      52013      83.1%         20   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   6265.690000   
1                   w   14135.68       14135.68   7553.490000   
2                   w   26017.15       26017.15  11247.560000   
3                   w       0.00           0.00   5637.052785   
4                   w    8968.74        8968.74   7057.950000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          6265.69          5136.86        1128.83                 0.0   
1          7553.49          4364.32        3189.17                 0.0   
2         11247.56          5982.85        5264.71                 0.0   
3          5637.05          5000.00         622.05                15.0   
4          7057.95          4031.26        3026.69                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0         0.0                      0.0     Oct-2019           570.81   
1         0.0                      0.0     Jun-2020           420.75   
2         0.0                      0.0     May-2020             0.00   
3         0.0                      0.0     Oct-2019          4741.09   
4         0.0                      0.0     Jun-2020             0.00   

  next_pym

### Clean up the `int_rate` column

When we're preparing a dataset for a machine learning model we typically want to represent don't want to leave any string values in our dataset --because it's hard to do math on words. 

Specifically, we have a column that is representing a numeric value, but currently doesn't have a numeric datatype. Lets look at the first 10 values of the `int_rate` column

In [93]:
# Look at the first 10 values of the int_rate column

df['int_rate'].head(10)

0      8.81%
1     12.98%
2     14.47%
3     17.97%
4     23.40%
5     26.31%
6     19.92%
7     23.40%
8     23.40%
9     11.80%
Name: int_rate, dtype: object

In [94]:
# Look at a specific value from the int_rate column
df['int_rate'][1]

' 12.98%'

Problems that we need to address with this column:

- String column that should be numeric
- Percent Sign `%` included with the number
- Leading space at the beginning of the string

However, we're not going to try and write exactly the right code to fix this column in one go. We're going to methodically build up to the code that will help us address these problems.


In [95]:
# Lets start with just fixing a single string.
# If we can fix one, we can usually fix all of them

int_rate = " 12.45%"

int_rate

' 12.45%'

In [96]:
int_rate.strip().strip('%')

'12.45'

In [97]:
int_rate

' 12.45%'

In [ ]:
# "Cast" the string value to a float

In [98]:
float(int_rate.strip().strip('%'))

12.45

### Write a function to make our solution reusable!

In [168]:
# Write a function that can do what we have written above to any 
# string that is passsed to it.

def int_rate_to_float(int_rate):
  return float(int_rate.strip().strip('%'))

In [100]:
# Test out our function by calling it on our example

int_rate_to_float(int_rate)

12.45

In [101]:
# is the data type correct?

type(int_rate_to_float(' 14.56%'))

float

### Apply our solution to every cell in a column

In [169]:
df['cleaned_int_rate'] = df['int_rate'].apply(int_rate_to_float)

df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN      18000        18000          18000.0   36 months   
1 NaN        NaN      18500        18500          18500.0   60 months   
2 NaN        NaN      32000        32000          32000.0   60 months   
3 NaN        NaN       5000         5000           5000.0   36 months   
4 NaN        NaN      13000        13000          13000.0   36 months   

  int_rate  installment grade sub_grade             emp_title emp_length  \
0    8.81%       570.81     A        A5                 Sever   < 1 year   
1   12.98%       420.75     B        B5  Supply Chain Analyst    5 years   
2   14.47%       752.41     C        C2               Teacher  10+ years   
3   17.97%       180.69     D        D1        Administrative    6 years   
4   23.40%       505.95     E        E1   Sale Representative    2 years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     40000.0     Source Verified  Dec-2018  Charged Off   
1           RENT     96500.0     Source Verified  Dec-2018      Current   
2       MORTGAGE    110000.0     Source Verified  Dec-2018      Current   
3       MORTGAGE     59280.0     Source Verified  Dec-2018   Fully Paid   
4       MORTGAGE     90000.0            Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN  debt_consolidation       Debt consolidation    100xx   
1          n  NaN   NaN         credit_card  Credit card refinancing    554xx   
2          n  NaN   NaN    home_improvement         Home improvement    152xx   
3          n  NaN   NaN  debt_consolidation       Debt consolidation    490xx   
4          n  NaN   NaN               other                    Other    191xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  12.81            0         Apr-2014               0   
1         MN  27.86            1         May-2009               0   
2         PA  26.11            0         Oct-2000               0   
3         MI  10.51            0         Apr-2011               0   
4         PA  39.73            0         May-2007               0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN        15        0   
1                     7.0                     NaN        19        0   
2                    24.0                     NaN        18        0   
3                     NaN                     NaN         8        0   
4                     NaN                     NaN        14        0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0      16147      35.2%         21                   w       0.00   
1      19000      62.1%         38                   w   14135.68   
2      14833        43%         32                   w   26017.15   
3       4599      19.1%         13                   w       0.00   
4      52013      83.1%         20                   w    8968.74   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   6265.690000          6265.69          5136.86   
1       14135.68   7553.490000          7553.49          4364.32   
2       26017.15  11247.560000         11247.56          5982.85   
3           0.00   5637.052785          5637.05          5000.00   
4        8968.74   7057.950000          7057.95          4031.26   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0        1128.83                 0.0         0.0                      0.0   
1        3189.17                 0.0         0.0                      0.0   
2        5264.71                 0.0         0.0                      0.0   
3         622.05                15.0         0.0                      0.0   
4        3026.69                 0.0         0.0                      0.0   

  last_pymnt_d  last_pymnt_amnt ne

In [ ]:
# What type of data is held in our new column?

# Float

# Look at the datatypes of the last 5 columns

df.dtypes.tail(5)

settlement_date           object
settlement_amount        float64
settlement_percentage    float64
settlement_term          float64
cleaned_int_rate         float64
dtype: object

## Challenge

We can create a new column with our cleaned values or overwrite the original, whatever we think best suits our needs. On your assignment you will take the same approach in trying to methodically build up the complexity of your code until you have a few lines that will work for any cell in a column. At that point you'll contain all of that functionality in a reusable function block and then use the `.apply()` function to... well... apply those changes to an entire column.

# [Objective](#pandas-apply) Modify and Create Columns using `.apply()`



## Overview

We're already seen one example of using the `.apply()` function to clean up a column. Lets see if we can do it again, but this time on a slightly more complicated use case.

Remember, the goal here is to write a function that will work correctly on any **individual** cell of a specific column. Then we can reuse that function on those individual cells of a dataframe column via the `.apply()` function.

Lets clean up the emp_title "Employment Title" column!

## Follow Along

First we'll try and diagnose how bad the problem is and what improvements we might be able to make.

In [103]:
# Look at the top 20 employment titles

df['emp_title'].value_counts().head(20)

Teacher                     2090
Manager                     1773
Registered Nurse             952
Driver                       924
RN                           726
Supervisor                   697
Sales                        580
Project Manager              526
General Manager              523
Office Manager               521
Owner                        420
Director                     402
Truck Driver                 387
Operations Manager           387
Nurse                        326
Engineer                     325
Sales Manager                304
manager                      301
Supervisor                   270
Administrative Assistant     269
Name: emp_title, dtype: int64

In [104]:
# How many different unique employment titles are there currently?

df['emp_title'].nunique()

43892

In [105]:
# How often is the employment_title null?

df['emp_title'].isnull().sum()

20947

What are some possible reasons as to why a person's employment title may have not been provided?

In [170]:
# Create some examples that represent the cases that we want to clean up

examples = ['owner', 'Supervisor', ' Project Manager', np.NaN]

In [171]:
# Write a function to clean up these use cases and increase uniformity.

def clean_emp_title(title):
  if isinstance(title, str):
    return title.strip().title()
  else:
    return "Unknown"

for example in examples:
  print(clean_emp_title(example))

Owner
Supervisor
Project Manager
Unknown


In [108]:
# list comprehensions can combine function calls and for loops over lists
# into one succinct and fairly readable single line of code.

[clean_emp_title(title) for title in examples]

['Owner', 'Supervisor', 'Project Manager', 'Unknown']

In [172]:
# We have a function that works as expected. Lets apply it to our column.
# This time we'll overwrite the original column

df['emp_title'] = df['emp_title'].apply(clean_emp_title)

We can use the same code as we did earlier to see how much progress was made.


In [111]:
# Look at the top 20 employment titles
df['emp_title'].value_counts().head(20)

Unknown               20947
Teacher                2557
Manager                2395
Registered Nurse       1418
Driver                 1258
Supervisor             1160
Truck Driver            920
Rn                      834
Office Manager          805
Sales                   803
General Manager         791
Project Manager         720
Owner                   625
Director                523
Operations Manager      518
Sales Manager           500
Police Officer          440
Nurse                   425
Technician              420
Engineer                412
Name: emp_title, dtype: int64

In [112]:
# How many different unique employment titles are there currently?

df['emp_title'].nunique()

34902

In [113]:
# How often is the employment_title null (NaN)?

df['emp_title'].isnull().sum()

0

## Challenge

Using the .apply() function isn't always about creating new columns on a dataframe, we can use it to clean up or modify existing columns as well. 

# [Objective](#dates-and-times) Work with Dates and Times with Pandas

## Overview

Pandas has its own datatype datatype that makes it extremely convenient to convert strings that are in standard date formates to datetime objects and then use those datetime objects to either create new features on a dataframe or work with the dataset in a timeseries fashion. 

This section will demonstrate how to take a column of date strings, convert it to a datetime object and then use the datetime formatting `.dt` to access specific parts of the date (year, month, day) to generate useful columns on a dataframe.

## Follow Along

### Work with Dates 

pandas documentation
- [to_datetime](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)
- [Time/Date Components](https://pandas.pydata.org/pandas-docs/stable/timeseries.html#time-date-components) "You can access these properties via the `.dt` accessor"

Many of the most useful date columns in this dataset have the suffix `_d` to indicate that they correspond to dates.

We'll use a list comprehension to print them out

In [114]:
[col for col in df if col.endswith('_d')]

['issue_d', 'last_pymnt_d', 'next_pymnt_d', 'last_credit_pull_d']

Lets look at the string format of the `issue_d` column

In [115]:
df['issue_d'].head()

0    Dec-2018
1    Dec-2018
2    Dec-2018
3    Dec-2018
4    Dec-2018
Name: issue_d, dtype: object

Because this string format %m-%y is a common datetime format, we can just let Pandas detect this format and translate it to the appropriate datetime object.

In [116]:
df['issue_d'].value_counts()

Oct-2018    46305
Nov-2018    41973
Dec-2018    40134
Name: issue_d, dtype: int64

Now we can see that the `issue_d` column has been changed to hold `datetime` objects.

Lets look at one of the cells specifically to see what a datetime object looks like:

In [117]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)

You can see how the month and year have been indicated by the strings that were contained in the column previously, and that the rest of the values have been inferred.

In [118]:
df['issue_d'].head()

0   2018-12-01
1   2018-12-01
2   2018-12-01
3   2018-12-01
4   2018-12-01
Name: issue_d, dtype: datetime64[ns]

We can use the `.dt` accessor to now grab specific parts of the datetime object. Lets grab just the year from the all of the cells in the `issue_d` column

In [119]:
df['issue_d'].dt.year

0         2018
1         2018
2         2018
3         2018
4         2018
          ... 
128407    2018
128408    2018
128409    2018
128410    2018
128411    2018
Name: issue_d, Length: 128412, dtype: int64

Now the month.

In [120]:
df['issue_d'].dt.month

0         12
1         12
2         12
3         12
4         12
          ..
128407    10
128408    10
128409    10
128410    10
128411    10
Name: issue_d, Length: 128412, dtype: int64

It's just that easy! Now, instead of printing them out, lets add these year and month values as new columns on our dataframe. Again, you'll have to scroll all the way over to the right in the table to see the new columns.

In [121]:
df['issue_year'] = df['issue_d'].dt.year
df['issue_month'] = df['issue_d'].dt.month

Because all of these dates come from Q4 of 2018, the `issue_d` column isn't all that interesting. Lets look at the `earliest_cr_line` column, which is also a string, but that could be converted to datetime format.

We're going to create a new column called `days_from_earliest_credit_to_issue`

It's a long column header, but think about how valuable this piece of information could be. This number will essentially indicate the length of a person's credit history and if that is correlated with repayment or other factors could be a valuable predictor!

In [122]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True)

In [123]:
df['earliest_cr_line'].head()

0   2014-04-01
1   2009-05-01
2   2000-10-01
3   2011-04-01
4   2007-05-01
Name: earliest_cr_line, dtype: datetime64[ns]

What we're about to do is so cool! Pandas' datetime format is so smart that we can simply use the subtraction operator `-` in order to calculate the amount of time between two dates. 

Think about everything that's going on under the hood in order to give us such straightforward syntax! Handling months of different lengths, leap years, etc. Pandas datetime objects are seriously powerful!

In [124]:
(df['issue_d'] - df['earliest_cr_line']).dt.days / 365

0          4.671233
1          9.591781
2         18.178082
3          7.673973
4         11.594521
            ...    
128407    33.189041
128408    36.276712
128409    12.342466
128410    10.005479
128411    12.090411
Length: 128412, dtype: float64

In [125]:
df['credit_length'] = (df['issue_d'] - df['earliest_cr_line']).dt.days / 365

df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN      18000        18000          18000.0   36 months   
1 NaN        NaN      18500        18500          18500.0   60 months   
2 NaN        NaN      32000        32000          32000.0   60 months   
3 NaN        NaN       5000         5000           5000.0   36 months   
4 NaN        NaN      13000        13000          13000.0   36 months   

  int_rate  installment grade sub_grade             emp_title emp_length  \
0    8.81%       570.81     A        A5                 Sever   < 1 year   
1   12.98%       420.75     B        B5  Supply Chain Analyst    5 years   
2   14.47%       752.41     C        C2               Teacher  10+ years   
3   17.97%       180.69     D        D1        Administrative    6 years   
4   23.40%       505.95     E        E1   Sale Representative    2 years   

  home_ownership  annual_inc verification_status    issue_d  loan_status  \
0           RENT     40000.0     Source Verified 2018-12-01  Charged Off   
1           RENT     96500.0     Source Verified 2018-12-01      Current   
2       MORTGAGE    110000.0     Source Verified 2018-12-01      Current   
3       MORTGAGE     59280.0     Source Verified 2018-12-01   Fully Paid   
4       MORTGAGE     90000.0            Verified 2018-12-01      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN  debt_consolidation       Debt consolidation    100xx   
1          n  NaN   NaN         credit_card  Credit card refinancing    554xx   
2          n  NaN   NaN    home_improvement         Home improvement    152xx   
3          n  NaN   NaN  debt_consolidation       Debt consolidation    490xx   
4          n  NaN   NaN               other                    Other    191xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  12.81            0       2014-04-01               0   
1         MN  27.86            1       2009-05-01               0   
2         PA  26.11            0       2000-10-01               0   
3         MI  10.51            0       2011-04-01               0   
4         PA  39.73            0       2007-05-01               0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN        15        0   
1                     7.0                     NaN        19        0   
2                    24.0                     NaN        18        0   
3                     NaN                     NaN         8        0   
4                     NaN                     NaN        14        0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0      16147      35.2%         21                   w       0.00   
1      19000      62.1%         38                   w   14135.68   
2      14833        43%         32                   w   26017.15   
3       4599      19.1%         13                   w       0.00   
4      52013      83.1%         20                   w    8968.74   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   6265.690000          6265.69          5136.86   
1       14135.68   7553.490000          7553.49          4364.32   
2       26017.15  11247.560000         11247.56          5982.85   
3           0.00   5637.052785          5637.05          5000.00   
4        8968.74   7057.950000          7057.95          4031.26   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0        1128.83                 0.0         0.0                      0.0   
1        3189.17                 0.0         0.0                      0.0   
2        5264.71                 0.0         0.0                      0.0   
3         622.05                15.0         0.0                      0.0   
4        3026.69                 0.0         0.0                      0.0   

  last_pymnt_d  last_pymnt_a

What's oldest credit history that was involved in Q4 2018? 

In [126]:
df['credit_length'].describe()

count    128412.000000
mean         16.054497
std           7.908317
min           3.084932
25%          11.093151
50%          14.427397
75%          19.846575
max          68.961644
Name: credit_length, dtype: float64

25,171 days is ~ 68.96 years of credit history!

## Challenge

Pandas' datetime format is so easy to work with that there's really no excuse for not using dates to make features on a dataframe! Get ready to practice more of this on your assignment.

Assignment:

In [191]:
#turning term column from str to int

def term_months_to_int(termdata):
  return int(termdata.strip().strip('months'))

# copying dataset to avoid reloads

newdf = df.copy()

# converting term values to ints from str

newdf['term'] = [term_months_to_int(termdata) for termdata in df['term']]

# function for creating new loan_status_is_great column

def loan_status_is_great_calc(loan_status):
  if loan_status == "Current" or loan_status == "Fully Paid":
    return 1
  else:
    return 0

# creating new column

newdf['loan_status_is_great'] = newdf['loan_status'].apply(loan_status_is_great_calc)

# adding last payment year and month values to dataframe

newdf['last_pymnt_d'] = pd.to_datetime(newdf['last_pymnt_d'], infer_datetime_format=True)
newdf['issue_year'] = newdf['last_pymnt_d'].dt.year
newdf['issue_month'] = newdf['last_pymnt_d'].dt.month

newdf.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  term int_rate  \
0 NaN        NaN      18000        18000          18000.0    36    8.81%   
1 NaN        NaN      18500        18500          18500.0    60   12.98%   
2 NaN        NaN      32000        32000          32000.0    60   14.47%   
3 NaN        NaN       5000         5000           5000.0    36   17.97%   
4 NaN        NaN      13000        13000          13000.0    36   23.40%   

   installment grade sub_grade             emp_title emp_length  \
0       570.81     A        A5                 Sever   < 1 year   
1       420.75     B        B5  Supply Chain Analyst    5 years   
2       752.41     C        C2               Teacher  10+ years   
3       180.69     D        D1        Administrative    6 years   
4       505.95     E        E1   Sale Representative    2 years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     40000.0     Source Verified  Dec-2018  Charged Off   
1           RENT     96500.0     Source Verified  Dec-2018      Current   
2       MORTGAGE    110000.0     Source Verified  Dec-2018      Current   
3       MORTGAGE     59280.0     Source Verified  Dec-2018   Fully Paid   
4       MORTGAGE     90000.0            Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN  debt_consolidation       Debt consolidation    100xx   
1          n  NaN   NaN         credit_card  Credit card refinancing    554xx   
2          n  NaN   NaN    home_improvement         Home improvement    152xx   
3          n  NaN   NaN  debt_consolidation       Debt consolidation    490xx   
4          n  NaN   NaN               other                    Other    191xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  12.81            0         Apr-2014               0   
1         MN  27.86            1         May-2009               0   
2         PA  26.11            0         Oct-2000               0   
3         MI  10.51            0         Apr-2011               0   
4         PA  39.73            0         May-2007               0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN        15        0   
1                     7.0                     NaN        19        0   
2                    24.0                     NaN        18        0   
3                     NaN                     NaN         8        0   
4                     NaN                     NaN        14        0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0      16147      35.2%         21                   w       0.00   
1      19000      62.1%         38                   w   14135.68   
2      14833        43%         32                   w   26017.15   
3       4599      19.1%         13                   w       0.00   
4      52013      83.1%         20                   w    8968.74   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   6265.690000          6265.69          5136.86   
1       14135.68   7553.490000          7553.49          4364.32   
2       26017.15  11247.560000         11247.56          5982.85   
3           0.00   5637.052785          5637.05          5000.00   
4        8968.74   7057.950000          7057.95          4031.26   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0        1128.83                 0.0         0.0                      0.0   
1        3189.17                 0.0         0.0                      0.0   
2        5264.71                 0.0         0.0                      0.0   
3         622.05                15.0         0.0                      0.0   
4        3026.69                 0.0         0.0                      0.0   

  last_pymnt_d  last_pymnt_amnt next_pymnt_d last_credit_pull_d  \
0  